### PR TITLE
Added possibility to store env vars as secrets

### PR DESCRIPTION
**Pull request by mroman-redhat**
Closes: #48

### CODE REVIEW

File: jupyterhub_singleuser_profiles/profiles.py [hunk 1]
```diff
@@ -75,7 +76,7 @@ def get_config_maps_matching_label(self, target_label='jupyterhub=singleuser-pro
     _LOGGER.info("Found these additional Config Maps: %s" % config_maps_list)
     return config_maps_list
 
-  def read_config_map(self, config_map_name, key_name="profiles"):
+  def read_config_map(self, config_map_name, key_name="profile"):
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/profiles.py:79]: Will this break loading of profiles? Maybe we should remove the default and expect every function to provide the `key_name` when calling this method

File: jupyterhub_singleuser_profiles/profiles.py [hunk 2]
```diff
@@ -86,6 +87,24 @@ def read_config_map(self, config_map_name, key_name="profiles"):
     config_map_yaml = yaml.load(config_map.data[key_name])
     return config_map_yaml
 
+  def read_secret(self, secret_name, key_name="profile"):
+    print("SECRET NAME", secret_name, key_name)
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/profiles.py:91]: Remove debug print

File: jupyterhub_singleuser_profiles/profiles.py (hunk 1 above)
Comment (mroman-redhat) [reply jupyterhub_singleuser_profiles/profiles.py:79]: All original calls of this function seem to use set key_name to "profile", so those should not break, and mine do not set it, and use the modified default. I feel like keeping the default value there is safer.

File: jupyterhub_singleuser_profiles/profiles.py [hunk 3]
```diff
@@ -101,21 +120,62 @@ def write_config_map(self, config_map_name, key_name, data):
...
+          _LOGGER.error("Exception when calling CoreV1Api->create_namespaced_secret: %s\n" % e)
+      else:
+        raise
+
   def update_user_profile_cm(self, username, data={}):
     
     user_cm = self.get_user_profile_cm(username)
+    user_cm_env = []
+    user_secret = self.get_user_secret(username)
+    user_secret['env'] = []
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/profiles.py:143]: Is there a reason why you are handling CM and Secret differently - i.e.

For CM you create a separate variable and `append` to it, then set it to `cm['env']`
For Sercet you `append` directly to the `secret['env']`

File: jupyterhub_singleuser_profiles/profiles.py [hunk 4]
```diff
@@ -101,21 +120,62 @@ def write_config_map(self, config_map_name, key_name, data):
...
     cm_key_name = "profile"
     for key, value in data.items():
       user_cm[key] = value
+    for var in user_cm['env']:
+      if var['type'] == "password":
+        user_secret['env'].append(var) 
+      else:
+        user_cm_env.append(var)
+    user_cm['env'] = user_cm_env
+    self.write_secret(cm_name, cm_key_name, user_secret) #not a typo, secret uses the same name as user cm
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/profiles.py:154]: Should we rename the `cm_name` to something like `resource_name`?

File: jupyterhub_singleuser_profiles/profiles.py [hunk 5]
```diff
@@ -101,21 +120,62 @@ def write_config_map(self, config_map_name, key_name, data):
...
     self.write_config_map(cm_name, cm_key_name, user_cm)
 
   def get_user_profile_cm(self, username):
     cm = self.read_config_map(_USER_CONFIG_MAP_TEMPLATE % escape(username), "profile")
+    secret = self.get_user_secret(username)
     if cm == {}:
-      return _DEFAULT_USER_CM
+      cm = _DEFAULT_USER_CM
+    for sec in secret['env']:
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/profiles.py:162]: You will need to check if the env already exists here, for the case where admin adds some CM env or secret env manually and it collides with one that already exists - see

Screenshot from 2021-03-01 16-09-04

File: jupyterhub_singleuser_profiles/profiles.py [hunk 6]
```diff
@@ -101,21 +120,62 @@ def write_config_map(self, config_map_name, key_name, data):
...
     else:
-      return cm
+      return secret
+  
+  def update_user_secret(self, username, data={}):
+    secret = self.get_user_secret(username)
+    secret_name = _USER_CONFIG_MAP_TEMPLATE % escape(username)
+    for key, value in data.items():
+      secret[key] = value
+    self.write_secret()
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/profiles.py:178]: What is this methof? How come it does not have any arguments?

File: jupyterhub_singleuser_profiles/profiles.py [hunk 7]
```diff
@@ -101,21 +120,62 @@ def write_config_map(self, config_map_name, key_name, data):
...
+
+  def get_user_secret(self, username):
+    secret = self.read_secret(_USER_CONFIG_MAP_TEMPLATE % escape(username))
+    if secret == {}:
+      return{"env":[]}
     else:
-      return cm
+      return secret
+  
+  def update_user_secret(self, username, data={}):
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/profiles.py:173]: Where is this used?

Comment (mroman-redhat) [reply jupyterhub_singleuser_profiles/profiles.py:173]: I decided not to use this method, since I never had a use case where I needed to change just a single secret value. I will delete it.

Comment (mroman-redhat) [reply jupyterhub_singleuser_profiles/profiles.py:173]: I decided not to use this method, since I had no use cases where I needed to change some particular secrets in the secret ConfigMap. I will delete it.

File: jupyterhub_singleuser_profiles/profiles.py (hunk 6 above)
Comment (mroman-redhat) [reply jupyterhub_singleuser_profiles/profiles.py:178]: Part of the unfinished update method, it does not run (but would throw an error)

**Comment by mroman-redhat** [comment]
Just a note for the future to keep in mind: We wanted to add a hide text button, but we do not have an icon for it.

File: jupyterhub_singleuser_profiles/profiles.py (hunk 3 above)
Comment (mroman-redhat) [reply jupyterhub_singleuser_profiles/profiles.py:142]: I am reading `user_cm['env']` and dividing it between what stays in `user_cm` and what goes into secrets. That's why I needed the extra variable.

File: jupyterhub_singleuser_profiles/profiles.py [hunk 8]
```diff
@@ -101,21 +119,57 @@ def write_config_map(self, config_map_name, key_name, data):
...
+    self.write_config_map(resource_name, cm_key_name, user_cm)
 
   def get_user_profile_cm(self, username):
     cm = self.read_config_map(_USER_CONFIG_MAP_TEMPLATE % escape(username), "profile")
+    secret = self.get_user_secret(username)
     if cm == {}:
-      return _DEFAULT_USER_CM
+      cm = _DEFAULT_USER_CM
+    for sec in secret['env']:
+      if os.environ.get(sec):
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/profiles.py:162]: What exactly are you trying to do here?

File: jupyterhub_singleuser_profiles/profiles.py [hunk 9]
```diff
@@ -101,21 +119,57 @@ def write_config_map(self, config_map_name, key_name, data):
...
+        continue
+      cm['env'].append(sec)
+    return cm
+
+  def get_user_secret(self, username):
+    secret = self.read_secret(_USER_CONFIG_MAP_TEMPLATE % escape(username))
+    if secret == {}:
+      return{"env":[]}
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/profiles.py:170]: The default value should be defined at the top with `_DEFAULT_USER_CM`...ale missing space..

File: jupyterhub_singleuser_profiles/ui/src/CustomElements/VarForm.js [hunk 10]
```diff
@@ -2,35 +2,37 @@ import React from 'react'
...
             key: props.var_key,
             value: props.value,
-            type: "",
+            type: props.text_type,
         }
     }
 
     handleKeyChange(event) {
-        this.setState({key: event.target.value}, () => {this.checkKey(this.state.key)})
+        this.setState({key: event.target.value},() => {this.checkKey(this.state.key)})
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/ui/src/CustomElements/VarForm.js:18]: Why did you remove the space?

File: jupyterhub_singleuser_profiles/ui/src/CustomElements/VarForm.js [hunk 11]
```diff
@@ -39,22 +41,28 @@ class VarForm extends React.Component {
     }
 
     enterVariable(event) {
-        this.setState({key: event.target.text}, () => {this.checkKey(this.state.key)})
+        this.setState({key: event.target.text})
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/ui/src/CustomElements/VarForm.js:44]: You the argument for setState is dictionary, you don't need to call it twice, just use both keys in the dictionary

File: jupyterhub_singleuser_profiles/ui/src/CustomElements/VarForm.js [hunk 12]
```diff
@@ -39,22 +41,28 @@ class VarForm extends React.Component {
...
+        this.setState({type: "password"})
+        document.getElementById(this.state.key + "check"+this.props.formIndex).checked = true
```
Comment (vpavlin) [inline jupyterhub_singleuser_profiles/ui/src/CustomElements/VarForm.js:46]: It would be good to have a function which will generate the `id` - this way you have the "check" string hardcoded all over the place
